packages/cli/src/utils/build.js

```diff
@@ -71,171 +70,396 @@ const requiredFiles = async ({ cwd, entryPoints }) => {
+}
+
+// Opens a zip file for writing. Returns an Archiver object.
+const openZip = (outputPath) => {
+  const output = fs.createWriteStream(outputPath);
+  const zip = archiver('zip', {
+    zlib: { level: Z_BEST_COMPRESSION }, // Sets the compression level.
   });
 
-  const nodeMajorVersion = semver.coerce(constants.LAMBDA_VERSION).major;
-
-  return (
-    filePath.endsWith('package.json') ||
-    filePath.endsWith('definition.json') ||
-    // include old async deasync versions so this runs seamlessly across node versions
-    filePath.endsWith(path.join('bin', 'linux-x64-node-10', 'deasync.node')) ||
-    filePath.endsWith(path.join('bin', 'linux-x64-node-12', 'deasync.node')) ||
-    filePath.endsWith(path.join('bin', 'linux-x64-node-14', 'deasync.node')) ||
-    filePath.endsWith(
-      // Special, for zapier-platform-legacy-scripting-runner
-      path.join('bin', `linux-x64-node-${nodeMajorVersion}`, 'deasync.node'),
-    ) ||
-    filePath.match(
-      path.sep === '\\' ? /aws-sdk\\apis\\.*\.json/ : /aws-sdk\/apis\/.*\.json/,
-    ) ||
-    matchesConfigInclude
-  );
+  const streamCompletePromise = new Promise((resolve, reject) => {
+    output.on('close', resolve);
+    zip.on('error', reject);
+  });
+
+  zip.finish = async () => {
+    // zip.finalize() doesn't return a promise, so here we create a
+    // zip.finish() function so the caller can await it.
+    // So callers: Use `await zip.finish()` and avoid zip.finalize().
+    zip.finalize();
+    await streamCompletePromise;
+  };
+
+  if (path.sep === '\\') {
+    // On Windows, patch zip.file() and zip.symlink() so they normalize the path
+    // separator to '/' because we're supposed to use '/' in a zip file
+    // regardless of the OS platform. Those are the only two methods we're
+    // currently using. If you wanted to call other zip.xxx methods, you should
+    // patch them here as well.
+    const origFileMethod = zip.file;
+    zip.file = (filepath, data) => {
+      filepath = path.normalize(filepath);
+      return origFileMethod.call(zip, filepath, data);
+    };
+    const origSymlinkMethod = zip.symlink;
+    zip.symlink = (name, target, mode) => {
+      name = path.normalize(name);
+      target = path.normalize(target);
+      return origSymlinkMethod.call(zip, name, target, mode);
+    };
+  }
+
+  zip.pipe(output);
+  return zip;
 };
 
-const writeZipFromPaths = (dir, zipPath, paths) => {
-  return new Promise((resolve, reject) => {
-    const output = fs.createWriteStream(zipPath);
-    const zip = archiver('zip', {
-      zlib: { level: Z_BEST_COMPRESSION },
-    });
+const looksLikeWorkspaceRoot = async (dir) => {
+  if (fs.existsSync(path.join(dir, 'pnpm-workspace.yaml'))) {
+    return true;
+  }
 
-    // listen for all archive data to be written
-    output.on('close', function () {
-      resolve();
-    });
+  const packageJsonPath = path.join(dir, 'package.json');
+  if (!fs.existsSync(packageJsonPath)) {
```

Comment:
```suggestion
  if (! await fse.pathExists(packageJsonPath)) {
```
For non-blocking IO performance